firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/CheckForNewReleaseClientTest.java

```diff
@@ -311,34 +290,66 @@ public void handleNewReleaseFromClient_whenNewReleaseIsSameAsInstalledAab_return
   }
 
   @Test
-  public void isInstalledRelease_whenCodeHashesEqual_returnsTrue() {
-    doReturn(TEST_CODEHASH_1).when(checkForNewReleaseClient).extractApkCodeHash(any());
-    assertTrue(checkForNewReleaseClient.isInstalledRelease(TEST_RELEASE_NEWER_APK));
+  public void iisSameAsInstalledRelease_whenApkHashesEqual_returnsTrue() {
+    doReturn(CURRENT_APK_HASH).when(checkForNewReleaseClient).extractApkHash(any());
+    assertTrue(
+        checkForNewReleaseClient.isSameAsInstalledRelease(getTestInstalledRelease().build()));
   }
 
   @Test
-  public void isInstalledRelease_whenCodeHashesNotEqual_returnsFalse() {
-    doReturn(TEST_CODEHASH_2).when(checkForNewReleaseClient).extractApkCodeHash(any());
-    assertFalse(checkForNewReleaseClient.isInstalledRelease(TEST_RELEASE_NEWER_APK));
+  public void isSameAsInstalledRelease_whenApkHashesNotEqual_returnsFalse() {
+    doReturn(CURRENT_APK_HASH).when(checkForNewReleaseClient).extractApkHash(any());
+    assertFalse(checkForNewReleaseClient.isSameAsInstalledRelease(getTestNewRelease().build()));
   }
 
   @Test
-  public void extractApkCodeHash_ifKeyInCachedCodeHashes_doesNotRecalculateZipHash() {
+  public void isSameAsInstalledRelease_ifApkHashNotPresent_fallsBackToExternalCodeHash() {
+    doReturn(CURRENT_APK_HASH).when(checkForNewReleaseClient).extractApkHash(any());
+    when(mockReleaseIdentifierStorage.getExternalCodeHash(any())).thenReturn(CURRENT_CODEHASH);
+
+    assertFalse(
+        checkForNewReleaseClient.isSameAsInstalledRelease(
+            getTestNewRelease().setApkHash("").build()));
+    verify(mockReleaseIdentifierStorage).getExternalCodeHash(CURRENT_APK_HASH);
+  }
+
+  @Test
+  public void extractApkHash_ifKeyInCachedApkHashes_doesNotRecalculateZipHash() {
 
     try (MockedStatic mockedReleaseIdentificationUtils =
         mockStatic(ReleaseIdentificationUtils.class)) {
       PackageInfo packageInfo =
           shadowPackageManager.getInternalMutablePackageInfo(
               ApplicationProvider.getApplicationContext().getPackageName());
       mockedReleaseIdentificationUtils
-          .when(() -> ReleaseIdentificationUtils.calculateApkInternalCodeHash(any()))
-          .thenReturn(TEST_CODEHASH_1);
+          .when(() -> ReleaseIdentificationUtils.calculateApkHash(any()))
+          .thenReturn(NEW_CODEHASH);
 
-      checkForNewReleaseClient.extractApkCodeHash(packageInfo);
-      checkForNewReleaseClient.extractApkCodeHash(packageInfo);
+      checkForNewReleaseClient.extractApkHash(packageInfo);
+      checkForNewReleaseClient.extractApkHash(packageInfo);
       // check that calculateApkInternalCodeHash is only called once
       mockedReleaseIdentificationUtils.verify(
-          () -> ReleaseIdentificationUtils.calculateApkInternalCodeHash(any()));
+          () -> ReleaseIdentificationUtils.calculateApkHash(any()));
     }
   }
+
+  private AppDistributionReleaseInternal.Builder getTestNewRelease() {
```

Comment:
I modified this test to use the pattern described in: https://testing.googleblog.com/2018/02/testing-on-toilet-cleanly-create-test.html